### PR TITLE
Ability for admins to remove user from internal chat channels

### PR DESCRIPTION
**Pull request by Veggier**
#9054  What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description
References #9183
The admins can remove users from internal chat channels. Also, I made the `Channel Name` as required when admins create the channel
## QA Instructions, Screenshots, Recordings
 `Channel Name` as required 
image

before remove
image

after press remove button
image

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [x] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/controllers/internal/chat_channels_controller.rb [hunk 1]
```diff
@@ -23,6 +23,14 @@ def update
       redirect_back(fallback_location: "/internal/chat_channels")
     end
 
+    def destroy
+      @chat_channel = ChatChannel.find(params[:id])
+      @chat_channel.remove_user(users_by_param)
+
+      flash.notice = "User has been remove"
```
Comment (jacobherrington) [inline app/controllers/internal/chat_channels_controller.rb:30]:
```suggestion
      flash.notice = "User has been removed"
```

File: app/controllers/internal/chat_channels_controller.rb [hunk 2]
```diff
@@ -23,6 +23,14 @@ def update
       redirect_back(fallback_location: "/internal/chat_channels")
     end
 
+    def destroy
```
Comment (mstruve) [inline app/controllers/internal/chat_channels_controller.rb:26]: What do you think about renaming this method to be a little more clear? As is, I would expect this method to delete the chat channel. I think a name such as `remove_user` might be more suitable and easier to understand.

Comment (Veggier) [reply app/controllers/internal/chat_channels_controller.rb:26]: Great suggestion, let me rename it.

Comment (Veggier) [reply app/controllers/internal/chat_channels_controller.rb]: @mstruve I finished the renaming, may you check with me

**Comment by Veggier** [comment]
> Looking great! Could you add a test for this as well? Even just a single happy path one would be great. A good place for that would be `internal/requests/chat_channel_spec.rb`

sorry, I don't know how to do that.

File: app/controllers/internal/chat_channels_controller.rb [hunk 3]
```diff
@@ -23,12 +23,22 @@ def update
...
+    end
+
     private
 
     def users_by_param
       User.where(username: chat_channel_params[:usernames_string].downcase.delete(" ").split(","))
     end
 
+    def user_by_param
+      User.where(username: chat_channel_params[:usernames_string])
```
Comment (Zhao-Andy) [inline app/controllers/internal/chat_channels_controller.rb:39]: This should be `find_by` so we only pass in a singular User record.

```suggestion
      User.find_by(username: chat_channel_params[:username_string])
```

File: app/views/internal/chat_channels/index.html.erb [hunk 4]
```diff
@@ -4,54 +4,55 @@
...
-
 <%= paginate @group_chat_channels %>
-
 <table class="table">
   <thead>
     <tr>
       <th scope="col">Channel Name</th>
       <th scope="col">Users</th>
       <th scope="col">Add Users</th>
+      <th scope="col">Remove Users</th>
```
Comment (Zhao-Andy) [inline app/views/internal/chat_channels/index.html.erb:34]: Let's make this singular :)

```suggestion
      <th scope="col">Remove User</th>
```

File: app/views/internal/chat_channels/index.html.erb [hunk 5]
```diff
@@ -4,54 +4,55 @@
...
-<%= paginate @group_chat_channels %>
+          <td>
+            <%= form_for [:internal, channel] do |f| %>
+              <%= f.text_field :usernames_string, placeholder: "Usernames to add" %>
+              <%= f.submit %>
+            <% end %>
+          </td>
+          <td>
+            <%= form_with url: remove_user_internal_chat_channel_url(channel), model: [:internal, channel], method: :delete, local: true do |f| %>
+              <%= f.text_field :usernames_string, placeholder: "Usernames to remove" %>
```
Comment (Zhao-Andy) [inline app/views/internal/chat_channels/index.html.erb:50]: I think we should pass the param as singular here, too:

```suggestion
        <%= f.text_field :username_string, placeholder: "Username to remove" %>
```

File: spec/requests/internal/chat_channel_spec.rb [hunk 6]
```diff
@@ -21,4 +21,14 @@
       expect(ChatChannel.last.pending_users).to include(user)
     end
   end
+
+  describe "DELETE /internal/chat_channels/:id/remove_user" do
+    it "removes the user from the chat channel" do
+      delete "/internal/chat_channels/#{user.id}/remove_user",
+             params: { chat_channel: { usernames_string: user.username.to_s } }
```
Comment (Zhao-Andy) [inline spec/requests/internal/chat_channel_spec.rb:28]: Noting to update this param to singular `username_string`.

Also `user.username` always returns a string so you wont need `.to_s`.

File: spec/requests/internal/chat_channel_spec.rb [hunk 7]
```diff
@@ -21,4 +21,14 @@
...
+    end
+
+    expect(user.chat_channel_memberships.count).to eq 0
+    expect(ChatChannel.last.pending_users).to include(user)
```
Comment (Zhao-Andy) [inline spec/requests/internal/chat_channel_spec.rb:32]: I'm a bit confused here; why would the user be a pending user after being removed?

Comment (Veggier) [reply spec/requests/internal/chat_channel_spec.rb]: This was my first time doing the test, I was confused of doing that. I will try to fix that.

Comment (Zhao-Andy) [reply spec/requests/internal/chat_channel_spec.rb]: No worries! Feel free to ask questions :)

Comment (Veggier) [reply spec/requests/internal/chat_channel_spec.rb]: @Zhao-Andy  I have updated the test and pass successfully. May you check for me.

File: spec/requests/internal/chat_channel_spec.rb [hunk 8]
```diff
@@ -21,4 +21,31 @@
...
+      expect(chat_channel.users.count).to eq 0
+
+      delete "/internal/chat_channels/#{no_user_chat_channel.id}/remove_user",
+             params: { chat_channel: { username_string: user.username } }
+      expect(no_user_chat_channel.users.count).to eq 0
+
+      delete "/internal/chat_channels/#{mult_users_chat_channel.id}/remove_user",
+             params: { chat_channel: { username_string: user.username } }
+      expect(mult_users_chat_channel.users.count).to eq 1
+      expect(user.chat_channel_memberships.count).to eq 0
```
Comment (Zhao-Andy) [inline spec/requests/internal/chat_channel_spec.rb:48]: Appreciate the extra tests here, but I don't think we need them. :) I think testing the fact that the route and action work properly is more than enough.

Comment (Veggier) [reply spec/requests/internal/chat_channel_spec.rb:48]: ok, I will delete them, and left with one test.

Comment (Veggier) [reply spec/requests/internal/chat_channel_spec.rb]: @Zhao-Andy finished.

**Pull request merged**